python_modules/libraries/dagster-openai/dagster_openai_tests/test_resources.py

```diff
@@ -0,0 +1,539 @@
+
+        assert mock_client.called
+        assert mock_wrapper.called
+
+    @graph_asset
+    def openai_asset():
+        return openai_op(model_version_op(), message_op())
+
+    result = (
+        Definitions(
+            assets=[openai_asset],
+            jobs=[
+                define_asset_job(
+                    name="openai_asset_job", selection=AssetSelection.assets(openai_asset)
+                )
+            ],
+            resources={
+                "openai_resource": OpenAIResource(api_key="xoxp-1234123412341234-12341234-1234")
+            },
+        )
+        .get_job_def("openai_asset_job")
+        .execute_in_process()
+    )
+
+    assert result.success
+
+
+@patch("dagster_openai.resources.OpenAIResource._wrap_with_usage_metadata")
+@patch("dagster.AssetExecutionContext", autospec=AssetExecutionContext)
+@patch("dagster_openai.resources.Client")
+def test_openai_resource_with_multi_asset(mock_client, mock_context, mock_wrapper):
+    @multi_asset(
+        outs={
+            "status": AssetOut(),
+            "result": AssetOut(),
+        },
+    )
+    def openai_multi_asset(openai_resource: OpenAIResource):
+        assert openai_resource
+
+        mock_context.assets_def.keys_by_output_name.keys.return_value = [
+            AssetKey("status"),
+            AssetKey("result"),
+        ]
+        mock_context.output_for_asset_key.return_value = "result"
+
+        # Test success when asset_key is provided
+        with openai_resource.get_client(
+            context=mock_context, asset_key=AssetKey("result")
+        ) as client:
+            client.chat.completions.create(
+                model="gpt-3.5-turbo", messages=[{"role": "user", "content": "Say this is a test"}]
+            )
+
+        assert mock_client.called
+        assert mock_wrapper.call_count == 3
+        mock_wrapper.assert_called_with(
+            api_endpoint_class=ANY,
+            context=mock_context,
+            output_name="result",
+        )
+
+        # Test failure when asset_key is not provided
+        with pytest.raises(DagsterInvariantViolationError):
+            with openai_resource.get_client(context=mock_context) as client:
+                client.chat.completions.create(
+                    model="gpt-3.5-turbo",
+                    messages=[{"role": "user", "content": "Say this is a test"}],
+                )
+        return None, None
+
+    result = (
+        Definitions(
+            assets=[openai_multi_asset],
+            jobs=[
+                define_asset_job(
+                    name="openai_multi_asset_job",
+                    selection=AssetSelection.assets(openai_multi_asset),
+                )
+            ],
+            resources={
+                "openai_resource": OpenAIResource(api_key="xoxp-1234123412341234-12341234-1234")
+            },
+        )
+        .get_job_def("openai_multi_asset_job")
+        .execute_in_process()
```

Comment:
same as [2]

Reply:
Done in [d3581d5](https://github.com/dagster-io/dagster/pull/19697/commits/d3581d571913914f9a18087bb3268503c53894ba)